### ISSUE TITLE
Improve accuracy of Y to L conversion

### ISSUE THREAD

**Issue by heckflosse** [Beep6581/RawTherapee#4338]
For low Y values the conversion from Y to L is inaccurate.

Here's a patch with test code. It converts L to Y and back again 100 times and shows before after values.
To get the output, just open a raw file in editor.

```diff
diff --git a/rtengine/rawimagesource.cc b/rtengine/rawimagesource.cc
index 04066d49..7299169f 100644
--- a/rtengine/rawimagesource.cc
+++ b/rtengine/rawimagesource.cc
@@ -4623,6 +4623,30 @@ void RawImageSource::getAutoExpHistogram (LUTu & histogram, int& histcompr)
 // Histogram MUST be 256 in size; gamma is applied, blackpoint and gain also
 void RawImageSource::getRAWHistogram (LUTu & histRedRaw, LUTu & histGreenRaw, LUTu & histBlueRaw)
 {
+    float a = 0;
+    float b = 0;
+    
+    for (float L = 0; L < 2.f; L += 0.1f) {
+        float newL = L;
+        float x,y,z;
+        for (int i = 0; i < 100; ++i) {
+            Color::Lab2XYZ(newL,a,b,x,y,z);
+            Color::XYZ2Lab(x,y,z,newL,a,b);
+        }
+        printf("Lold = %.9f\tLnew = %.9f\n",L,newL);
+    }
+
+    for (float L = 2.f; L < 33000.f; L *= 2.f) {
+        float newL = L;
+        float x,y,z;
+        for (int i = 0; i < 100; ++i) {
+            Color::Lab2XYZ(newL,a,b,x,y,z);
+            Color::XYZ2Lab(x,y,z,newL,a,b);
+        }
+        printf("Lold = %.9f\tLnew = %.9f\n",L,newL);
+    }
+
+    std::cout << std::endl;
 //    BENCHFUN
     histRedRaw.clear();
     histGreenRaw.clear();

```

And here's the output:

```
Lold = 0.000000000      Lnew = 0.044082642
Lold = 0.100000001      Lnew = 0.144088745
Lold = 0.200000003      Lnew = 0.244094849
Lold = 0.300000012      Lnew = 0.344100952
Lold = 0.400000006      Lnew = 0.444107056
Lold = 0.500000000      Lnew = 0.544113159
Lold = 0.600000024      Lnew = 0.644119263
Lold = 0.700000048      Lnew = 0.700317383
Lold = 0.800000072      Lnew = 0.800323486
Lold = 0.900000095      Lnew = 0.900329590
Lold = 1.000000119      Lnew = 1.000335693
Lold = 1.100000143      Lnew = 1.100341797
Lold = 1.200000167      Lnew = 1.200347900
Lold = 1.300000191      Lnew = 1.300354004
Lold = 1.400000215      Lnew = 1.400360107
Lold = 1.500000238      Lnew = 1.500366211
Lold = 1.600000262      Lnew = 1.600372314
Lold = 1.700000286      Lnew = 1.699935913
Lold = 1.800000310      Lnew = 1.799942017
Lold = 1.900000334      Lnew = 1.899948120
Lold = 2.000000000      Lnew = 1.999954224
Lold = 4.000000000      Lnew = 4.000076294
Lold = 8.000000000      Lnew = 7.999877930
Lold = 16.000000000     Lnew = 16.000366211
Lold = 32.000000000     Lnew = 32.000015259
Lold = 64.000000000     Lnew = 64.044006348
Lold = 128.000000000    Lnew = 128.000122070
Lold = 256.000000000    Lnew = 255.999969482
Lold = 512.000000000    Lnew = 512.000122070
Lold = 1024.000000000   Lnew = 1023.999938965
Lold = 2048.000000000   Lnew = 2048.000000000
Lold = 4096.000000000   Lnew = 4099.652343750
Lold = 8192.000000000   Lnew = 8195.708007812
Lold = 16384.000000000  Lnew = 16387.683593750
Lold = 32768.000000000  Lnew = 32771.742187500
```

**Comment by heckflosse** [comment]
This patch improves accuracy:

```diff
diff --git a/rtengine/color.cc b/rtengine/color.cc
index 68fd467e..e88aa708 100644
--- a/rtengine/color.cc
+++ b/rtengine/color.cc
@@ -34,6 +34,7 @@ extern const Settings* settings;
 
 cmsToneCurve* Color::linearGammaTRC;
 LUTf Color::cachef;
+LUTf Color::cachefy;
 LUTf Color::gamma2curve;
 
 LUTf Color::gammatab;
@@ -137,6 +138,7 @@ void Color::init ()
     constexpr auto maxindex = 65536;
 
     cachef(maxindex, LUT_CLIP_BELOW);
+    cachefy(maxindex, LUT_CLIP_BELOW);
     gammatab(maxindex, 0);
     gammatabThumb(maxindex, 0);
 
@@ -180,6 +182,23 @@ void Color::init ()
         }
 #ifdef _OPENMP
         #pragma omp section
+#endif
+        {
+            int i = 0;
+            int epsmaxint = eps_max;
+
+            for (; i <= epsmaxint; i++)
+            {
+                cachefy[i] = 327.68 * (kappa * i / MAXVALF);
+            }
+
+            for(; i < maxindex; i++)
+            {
+                cachefy[i] = 327.68 * (116.0 * std::cbrt((double)i / MAXVALF) - 16.0);
+            }
+        }
+#ifdef _OPENMP
+        #pragma omp section
 #endif
         {
             for (int i = 0; i < maxindex; i++)
@@ -1799,7 +1818,8 @@ void Color::XYZ2Lab(float X, float Y, float Z, float &L, float &a, float &b)
     fy = (y <= 65535.0f ? cachef[y] : (327.68f * xcbrtf(y / MAXVALF)));
     fz = (z <= 65535.0f ? cachef[z] : (327.68f * xcbrtf(z / MAXVALF)));
 
-    L = (116.0f *  fy - 5242.88f); //5242.88=16.0*327.68;
+    L = (y <= 65535.0f ? cachefy[y] : 327.68f * (116.f * xcbrtf(y / MAXVALF) - 16.f));
+//    L = (116.0f *  fy - 5242.88f); //5242.88=16.0*327.68;
     a = (500.0f * (fx - fy) );
     b = (200.0f * (fy - fz) );
 }
diff --git a/rtengine/color.h b/rtengine/color.h
index d56b30e5..083761c6 100644
--- a/rtengine/color.h
+++ b/rtengine/color.h
@@ -129,6 +129,7 @@ public:
     static cmsToneCurve* linearGammaTRC;
 
     static LUTf cachef;
+    static LUTf cachefy;
     static LUTf gamma2curve;
 
     // look-up tables for the standard srgb gamma and its inverse (filled by init())

```

Output:

```
Lold = 0.000000000      Lnew = 0.000000000
Lold = 0.100000001      Lnew = 0.100000001
Lold = 0.200000003      Lnew = 0.200000003
Lold = 0.300000012      Lnew = 0.300000072
Lold = 0.400000006      Lnew = 0.400000006
Lold = 0.500000000      Lnew = 0.500000119
Lold = 0.600000024      Lnew = 0.600000143
Lold = 0.700000048      Lnew = 0.700000167
Lold = 0.800000072      Lnew = 0.800000310
Lold = 0.900000095      Lnew = 0.900000155
Lold = 1.000000119      Lnew = 1.000000238
Lold = 1.100000143      Lnew = 1.100000143
Lold = 1.200000167      Lnew = 1.200000286
Lold = 1.300000191      Lnew = 1.300000191
Lold = 1.400000215      Lnew = 1.400000334
Lold = 1.500000238      Lnew = 1.500000238
Lold = 1.600000262      Lnew = 1.600000620
Lold = 1.700000286      Lnew = 1.700000644
Lold = 1.800000310      Lnew = 1.800000310
Lold = 1.900000334      Lnew = 1.900000334
Lold = 2.000000000      Lnew = 2.000000477
Lold = 4.000000000      Lnew = 4.000000954
Lold = 8.000000000      Lnew = 8.000001907
Lold = 16.000000000     Lnew = 16.000000000
Lold = 32.000000000     Lnew = 32.000000000
Lold = 64.000000000     Lnew = 64.000000000
Lold = 128.000000000    Lnew = 128.000000000
Lold = 256.000000000    Lnew = 256.000000000
Lold = 512.000000000    Lnew = 512.000000000
Lold = 1024.000000000   Lnew = 1024.000000000
Lold = 2048.000000000   Lnew = 2048.000000000
Lold = 4096.000000000   Lnew = 4099.668457031
Lold = 8192.000000000   Lnew = 8195.656250000
Lold = 16384.000000000  Lnew = 16387.679687500
Lold = 32768.000000000  Lnew = 32771.613281250
```

**Comment by heckflosse** [comment]
Here's an improved patch which also increases accuracy in the upper range:

```diff
diff --git a/rtengine/color.cc b/rtengine/color.cc
index 68fd467e..d224eaca 100644
--- a/rtengine/color.cc
+++ b/rtengine/color.cc
@@ -34,6 +34,7 @@ extern const Settings* settings;
 
 cmsToneCurve* Color::linearGammaTRC;
 LUTf Color::cachef;
+LUTf Color::cachefy;
 LUTf Color::gamma2curve;
 
 LUTf Color::gammatab;
@@ -137,6 +138,7 @@ void Color::init ()
     constexpr auto maxindex = 65536;
 
     cachef(maxindex, LUT_CLIP_BELOW);
+    cachefy(maxindex, LUT_CLIP_BELOW);
     gammatab(maxindex, 0);
     gammatabThumb(maxindex, 0);
 
@@ -180,6 +182,23 @@ void Color::init ()
         }
 #ifdef _OPENMP
         #pragma omp section
+#endif
+        {
+            int i = 0;
+            int epsmaxint = eps_max;
+
+            for (; i <= epsmaxint; i++)
+            {
+                cachefy[i] = 327.68 * (kappa * i / MAXVALF);
+            }
+
+            for(; i < maxindex; i++)
+            {
+                cachefy[i] = 327.68 * (116.0 * std::cbrt((double)i / MAXVALF) - 16.0);
+            }
+        }
+#ifdef _OPENMP
+        #pragma omp section
 #endif
         {
             for (int i = 0; i < maxindex; i++)
@@ -1683,7 +1702,7 @@ void Color::Lab2XYZ(float L, float a, float b, float &x, float &y, float &z)
     float LL = L / 327.68f;
     float aa = a / 327.68f;
     float bb = b / 327.68f;
-    float fy = (0.00862069f * LL) + 0.137932f; // (L+16)/116
+    float fy = (0.00862068965517241379310344827586f * LL) + 0.13793103448275862068965517241379f; // (L+16)/116
     float fx = (0.002f * aa) + fy;
     float fz = fy - (0.005f * bb);
     x = 65535.0f * f2xyz(fx) * D50x;
@@ -1799,7 +1818,8 @@ void Color::XYZ2Lab(float X, float Y, float Z, float &L, float &a, float &b)
     fy = (y <= 65535.0f ? cachef[y] : (327.68f * xcbrtf(y / MAXVALF)));
     fz = (z <= 65535.0f ? cachef[z] : (327.68f * xcbrtf(z / MAXVALF)));
 
-    L = (116.0f *  fy - 5242.88f); //5242.88=16.0*327.68;
+    L = (y <= 65535.0f ? cachefy[y] : 327.68f * (116.f * xcbrtf(y / MAXVALF) - 16.f));
+//    L = (116.0f *  fy - 5242.88f); //5242.88=16.0*327.68;
     a = (500.0f * (fx - fy) );
     b = (200.0f * (fy - fz) );
 }
diff --git a/rtengine/color.h b/rtengine/color.h
index d56b30e5..083761c6 100644
--- a/rtengine/color.h
+++ b/rtengine/color.h
@@ -129,6 +129,7 @@ public:
     static cmsToneCurve* linearGammaTRC;
 
     static LUTf cachef;
+    static LUTf cachefy;
     static LUTf gamma2curve;
 
     // look-up tables for the standard srgb gamma and its inverse (filled by init())

```

Output:

```
Lold = 0.000000000      Lnew = 0.000000000
Lold = 0.100000001      Lnew = 0.100000001
Lold = 0.200000003      Lnew = 0.200000003
Lold = 0.300000012      Lnew = 0.300000072
Lold = 0.400000006      Lnew = 0.400000006
Lold = 0.500000000      Lnew = 0.500000119
Lold = 0.600000024      Lnew = 0.600000143
Lold = 0.700000048      Lnew = 0.700000167
Lold = 0.800000072      Lnew = 0.800000310
Lold = 0.900000095      Lnew = 0.900000155
Lold = 1.000000119      Lnew = 1.000000238
Lold = 1.100000143      Lnew = 1.100000143
Lold = 1.200000167      Lnew = 1.200000286
Lold = 1.300000191      Lnew = 1.300000191
Lold = 1.400000215      Lnew = 1.400000334
Lold = 1.500000238      Lnew = 1.500000238
Lold = 1.600000262      Lnew = 1.600000620
Lold = 1.700000286      Lnew = 1.700000644
Lold = 1.800000310      Lnew = 1.800000310
Lold = 1.900000334      Lnew = 1.900000334
Lold = 2.000000000      Lnew = 2.000000477
Lold = 4.000000000      Lnew = 4.000000954
Lold = 8.000000000      Lnew = 8.000001907
Lold = 16.000000000     Lnew = 16.000000000
Lold = 32.000000000     Lnew = 32.000000000
Lold = 64.000000000     Lnew = 64.000000000
Lold = 128.000000000    Lnew = 128.000000000
Lold = 256.000000000    Lnew = 256.000000000
Lold = 512.000000000    Lnew = 512.000000000
Lold = 1024.000000000   Lnew = 1024.000000000
Lold = 2048.000000000   Lnew = 2048.000000000
Lold = 4096.000000000   Lnew = 4096.000000000
Lold = 8192.000000000   Lnew = 8192.000000000
Lold = 16384.000000000  Lnew = 16383.999023438
Lold = 32768.000000000  Lnew = 32768.000000000
```

**Comment by Floessie** [comment]
Ingo 🔆 Weyrich. Yay! 👍

**Comment by heckflosse** [comment]
@Floessie Unfortunately it doesn't fix #4334 when clamping low L values to 0.f :( Still investigating...
Nevertheless I will prepare a patch which uses the new method with improved accuracy at all places in code I can find

**Comment by heckflosse** [comment]
@Floessie @agriggio Code reviews are welcome :)
